File: solutions/ruby/phone-number/2/phone_number.rb
module BookKeeping
  VERSION = 1
end

# PhoneNumber
module PhoneNumber
  module_function

  def clean(n)
    PhoneSanitizer.new(n).to_s
  end

  # PhoneSanitizer
  class PhoneSanitizer
    def initialize(number)
      @number = only_digits(number)
      @number = remove_international_code(@number)
    end

    def to_s
      sane_length? && sane_form? ? @number : nil
    end

    private

    def sane_length?
      @number.length == 10
    end

    def sane_form?
      @number =~ /^[2-9]\d\d[2-9]\d{6}$/
    end

    def remove_international_code(number)
      number.length == 11 && number[0] == '1' ? number[1..-1] : number
    end

    def only_digits(number)
      number.gsub(/\D/, '')
    end
  end
end

File: solutions/lua/phone-number/2/phone-number.lua
return {
  new = function(_, number)
    number = number:gsub("^1", ""):gsub("%D", "")
    number = number:len() == 10 and number or "0000000000"
    return setmetatable({ number = number, areaCode = function(self) return self.number:sub(1, 3) end },
      { __tostring = function(self) return self.number:gsub('(...)(...)(....)', "(%1) %2-%3") end })
  end
}
